packages/datagrid/src/selectionmodel.ts

```diff
@@ -0,0 +1,337 @@
+/*-----------------------------------------------------------------------------
+| Copyright (c) 2014-2019, PhosphorJS Contributors
+|
+| Distributed under the terms of the BSD 3-Clause License.
+|
+| The full license is in the file LICENSE, distributed with this software.
+|----------------------------------------------------------------------------*/
+import {
+  IIterator, some
+} from '@phosphor/algorithm';
+
+import {
+  ISignal, Signal
+} from '@phosphor/signaling';
+
+import {
+  DataModel
+} from './datamodel';
+
+
+/**
+ * A base class for creating data grid selection models.
+ *
+ * #### Notes
+ * If the predefined selection models are insufficient for a particular
+ * use case, a custom model can be defined which derives from this class.
+ */
+export
+abstract class SelectionModel {
+  /**
+   * Construct a new selection model.
+   *
+   * @param options - The options for initializing the model
+   */
+  constructor(options: SelectionModel.IOptions) {
+    this.model = options.model;
+    this._selectionMode = options.selectionMode || 'cell';
+    this.model.changed.connect(this.onModelChanged, this);
```

Comment:
When a selection model gets set to `null` in the data grid, does this connection need to be cleaned up?

Reply:
Good question. Since a single selection model can be shared among multiple grids (to have linked selections), it's not clear who the owner of a selection model really is. This is why I didn't make the selection model an `IDisposable`.